packages/utils/src/dal/mikro-orm/db-error-mapper.ts

```diff
@@ -0,0 +1,82 @@
+import {
+  ForeignKeyConstraintViolationException,
+  InvalidFieldNameException,
+  NotFoundError,
+  NotNullConstraintViolationException,
+  UniqueConstraintViolationException,
+} from "@mikro-orm/core"
+import { MedusaError, upperCaseFirst } from "../../common"
+
+export const dbErrorMapper = (err: Error) => {
+  if (err instanceof NotFoundError) {
+    throw new MedusaError(MedusaError.Types.NOT_FOUND, err.message)
+  }
+
+  if (err instanceof UniqueConstraintViolationException) {
+    const info = getConstraintInfo(err)
+    if (!info) {
+      throw err
+    }
+
+    throw new MedusaError(
+      MedusaError.Types.INVALID_DATA,
+      `${upperCaseFirst(info.table)} with ${info.keys
+        .map((key, i) => `${key}: ${info.values[i]}`)
+        .join(", ")} already exists.`
+    )
+  }
+
+  if (err instanceof NotNullConstraintViolationException) {
+    throw new MedusaError(
+      MedusaError.Types.INVALID_DATA,
+      `Cannot set field '${(err as any).column}' of ${upperCaseFirst(
+        (err as any).table
+      )} to null`
+    )
+  }
+
+  if (err instanceof InvalidFieldNameException) {
+    const userFriendlyMessage = err.message.match(/(column.*)/)?.[0]
+    throw new MedusaError(
+      MedusaError.Types.INVALID_DATA,
+      userFriendlyMessage ?? err.message
+    )
+  }
+
+  if (err instanceof ForeignKeyConstraintViolationException) {
+    const info = getConstraintInfo(err)
+    throw new MedusaError(
+      MedusaError.Types.INVALID_DATA,
+      `You tried to set relationship ${info?.keys.map(
+        (key, i) => `${key}: ${info.values[i]}`
+      )}, but such entity does not exist`
+    )
+  }
+
+  throw err
+}
+
+const getConstraintInfo = (err: any) => {
+  const detail = err.detail as string
+  if (!detail) {
+    return null
+  }
+
+  const [keys, values] = detail.match(/\([^\(.]*\)/g) || []
+
+  if (!keys || !values) {
+    return null
+  }
+
+  return {
```

Comment:
Can we also return the name of the constraint (err.constraint)? I believe adding it to the message in parentheses would be helpful.

Reply:
@carlos-r-l-rodrigues I think that might be too much of an implementation detail that we leak to the FE. What would be nice though is to have it as part of MedusaError in some form and then we plug that into some error logging/tracing before responding to the end user, but I think we can tackle that later when we integrate better with some tracing library

Reply:
Here is the DAL utils, so I expect the backend to have access to these properties.
it is alright to throw from here directly to the front end, but we need to have access to the details on the backend, even to handle errors differently based on the constraint name.

Reply:
@carlos-r-l-rodrigues yeah definitely, whichi s why I think we should add the constraint on MedusaError as a separate field, but not as part of the message. I suggest we just handle that separately though